sql/memory.go

```diff
@@ -0,0 +1,198 @@
+type RowsCache interface {
+	// Add a new row to the cache. If there is no memory available, it will try to
+	// free some memory. If after that there is still no memory available, it
+	// will return an error and erase all the content of the cache.
+	Add(Row) error
+	// Get all rows.
+	Get() []Row
+}
+
+// ErrNoMemoryAvailable is returned when there is no more available memory.
+var ErrNoMemoryAvailable = errors.NewKind("no memory available")
+
+const maxMemoryKey = "MAX_MEMORY"
+
+const (
+	b   = 1
+	kib = 1024 * b
+	mib = 1024 * kib
+)
+
+var maxMemory = func() uint64 {
+	val := os.Getenv(maxMemoryKey)
+	var v uint64
+	if val != "" {
+		var err error
+		v, err = strconv.ParseUint(val, 10, 64)
+		if err != nil {
+			panic("MAX_MEMORY environment variable must be a number, but got: " + val)
+		}
+	}
+
+	return v * uint64(mib)
+}()
+
+// Reporter is a component that gives information about the memory usage.
+type Reporter interface {
+	// MaxMemory returns the maximum number of memory allowed in bytes.
+	MaxMemory() uint64
+	// UsedMemory returns the memory in use in bytes.
+	UsedMemory() uint64
+}
+
+// ProcessMemory is a reporter for the memory used by the process and the
+// maximum amount of memory allowed controlled by the MAX_MEMORY environment
+// variable.
+var ProcessMemory Reporter = new(processReporter)
+
+type processReporter struct{}
+
+func (processReporter) UsedMemory() uint64 {
+	var s runtime.MemStats
+	runtime.ReadMemStats(&s)
+	return s.HeapInuse + s.StackInuse
+}
+
+func (processReporter) MaxMemory() uint64 { return maxMemory }
+
+// HasAvailableMemory reports whether more memory is available to the program if
+// it hasn't reached the max memory limit.
+func HasAvailableMemory(r Reporter) bool {
+	maxMemory := r.MaxMemory()
+	if maxMemory == 0 {
+		return true
+	}
+
+	return r.UsedMemory() < maxMemory
+}
+
+// MemoryManager is in charge of keeping track and managing all the components that operate
+// in memory. There should only be one instance of a memory manager running at the
+// same time in each process.
+type MemoryManager struct {
+	mu       sync.RWMutex
+	reporter Reporter
+	caches   map[uint64]Disposable
+	token    uint64
+}
+
+// NewMemoryManager creates a new manager with the given memory reporter. If nil is given,
+// then the Process reporter will be used by default.
+func NewMemoryManager(r Reporter) *MemoryManager {
+	if r == nil {
+		r = ProcessMemory
+	}
+
+	return &MemoryManager{
+		reporter: r,
+		caches:   make(map[uint64]Disposable),
+	}
+}
+
+// HasAvailable reports whether the memory manager has any available memory.
+func (m *MemoryManager) HasAvailable() bool {
+	return HasAvailableMemory(m.reporter)
+}
+
+// DisposeFunc is a function to completely erase a cache and remove it from the manager.
+type DisposeFunc func()
+
+// NewLRUCache returns an empty LRU cache and a function to dispose it when it's
+// no longer needed.
+func (m *MemoryManager) NewLRUCache(size uint) (KeyValueCache, DisposeFunc) {
```

Comment:
This is intended to be used per query or while the SQL engine is running?

Reply:
the same question for all the cache implementations

Reply:
While it's running.